src/custom-element.js

```diff
@@ -123,92 +123,103 @@ export function stubElements(win) {
+  if ((!inputLayout || inputLayout == Layout.FIXED ||
+          inputLayout == Layout.FIXED_HEIGHT) &&
+      (!inputWidth || !inputHeight) && hasNaturalDimensions(element.tagName)) {
+    // Default width and height: handle elements that do not specify a
+    // width/height and are defined to have natural browser dimensions.
     const dimensions = getNaturalDimensions(element.tagName);
-    if (layoutAttr != Layout.FIXED_HEIGHT) {
-      widthAttr = widthAttr || dimensions.width;
-    }
-    heightAttr = heightAttr || dimensions.height;
+    width = inputWidth || inputLayout == Layout.FIXED_HEIGHT ? inputWidth :
+        dimensions.width;
+    height = inputHeight || dimensions.height;
+  } else {
+    width = inputWidth;
+    height = inputHeight;
   }
 
-  let layout;
-  if (layoutAttr) {
-    // TODO(dvoytenko): show error state visually in the dev mode, e.g.
-    // red background + error message.
-    layout = parseLayout(layoutAttr.trim());
-    if (!layout) {
-      throw new Error('Unknown layout: ' + layoutAttr);
-    }
-  } else if (widthAttr || heightAttr) {
-    if (!widthAttr || widthAttr == 'auto') {
-      layout = Layout.FIXED_HEIGHT;
+  // Calculate effective layout.
+  if (inputLayout) {
+    // Layout specified directly.
+    layout = inputLayout;
+  } else {
+    // Default layout value.
+    if (width || height) {
+      if (!width || width == 'auto') {
+        layout = Layout.FIXED_HEIGHT;
+      } else {
+        layout = sizesAttr ? Layout.RESPONSIVE : Layout.FIXED;
+      }
     } else {
-      layout = sizesAttr ? Layout.RESPONSIVE : Layout.FIXED;
+      layout = Layout.CONTAINER;
     }
-  } else {
-    layout = Layout.CONTAINER;
-  }
-  element.classList.add(getLayoutClass(layout));
-  if (isLayoutSizeDefined(layout)) {
-    element.classList.add('-amp-layout-size-defined');
   }
 
+  // Verify layout attributes.
   if (layout == Layout.FIXED || layout == Layout.FIXED_HEIGHT ||
           layout == Layout.RESPONSIVE) {
-    let width = 0;
+    // These layouts must specify height and, in most cases, width.
+    assert(height, 'Expected height to be available: %s', heightAttr);
     if (layout == Layout.FIXED_HEIGHT) {
-      if (widthAttr && widthAttr != 'auto') {
-        throw new Error('Expected width to be either absent or equal "auto" ' +
-            'for fixed-height layout: ' + widthAttr);
-      }
+      assert(!width || width == 'auto',
+          'Expected width to be either absent or equal "auto" ' +
+          'for fixed-height layout: %s', width);//XXX: widthAttr
     } else {
-      width = parseLength(widthAttr);
-      if (!width) {
-        throw new Error('Expected width to be available and be an ' +
-            'integer/length value: ' + widthAttr);
-      }
-    }
-    const height = parseLength(heightAttr);
-    if (!height) {
-      throw new Error('Expected height to be available and be an ' +
-          'integer/length value: ' + heightAttr);
+      assert(width && width != 'auto',
```

Comment:
Have you considered putting the condition onto the block more directly? Right now I have to read the guard in l. 178, then negate the condition in 182 and subtract it. This could just read
if (layout == Layout.FIXED || layout == Layout.RESPONSIVE) {
  assert(width && ...
}
So basically I think getting rid of the guard and making each condition seperate is nicer here, because most conditions only involve one or two layouts. OK, this is a nitpick.